datafusion/physical-expr/src/expressions/binary.rs

```diff
@@ -430,17 +431,17 @@ fn string_concat(left: ArrayRef, right: ArrayRef) -> Result<ArrayRef> {
         scalar_value => scalar_value.into_array(left.clone().len()),
     };
     let ignore_null_array = ignore_null.as_any().downcast_ref::<StringArray>().unwrap();
-    let result = (0..ignore_null_array.len())
+    let index_array = (0..ignore_null_array.len())
         .into_iter()
         .map(|index| {
             if left.is_null(index) || right.is_null(index) {
                 None
             } else {
-                Some(ignore_null_array.value(index))
+                Some(index as u32)
             }
         })
-        .collect::<StringArray>();
-
+        .collect::<UInt32Array>();
+    let result = take(ignore_null_array, &index_array, None)?;
```

Comment:
`take` still potentially does a copy into a new array, though it is more optimized than the simple loop that was here

Reply:
I think we can optimize the whole string concat process to avoid this value copying, something like:

- original process: build-in `concat` ignoring `NULL` -> generate validity array or array of indexes -> take valid value from `concat` output array according to bitmap or indexes
- opmized process: concat two input string array well handled with `NULL`, no value copy any more.

Reply:
AIUC, the suggestion of @alamb might be roughly:


* Create the validity array based on `left_bitmap && right_bitmap` (this is an fast operation, as it can be performed on 64 bits at once). A fast path could be implemented for the null count being equal to 0 or being equal to length of the array.

* Compute the concatenation based on left + right side (without needing taking into account null values).

We should add some benches to see the improvements this brings.

In the arrow-rs kernels you could find some examples of this strategy implemented, but usually this brings 2-10x improvements, depending on data type.

In arrow2 this strategy is applied quite a bit more consistently, see for example the `substring` kernel: https://github.com/jorgecarleitao/arrow2/blob/main/src/compute/substring.rs#L66

Reply:
@Dandandan Thanks for you information, I' ll learn it later and try some benches on it

Reply:
Filed https://github.com/apache/arrow-rs/issues/1540 to add an optimized implementation to arrow-rs